Sources/CollectionDiffing.swift

```diff
@@ -0,0 +1,560 @@
+import Foundation
+import Result
+
+/// Represents a snapshot of a collection.
+///
+/// The changeset associated with the first snapshot received by any given observation
+/// is generally irrelevant. Observations should special case the first snapshot as a
+/// complete replacement, or whatever semantic that fits their purpose.
+///
+/// `Snapshot` does not include a reference to the previous snapshot. If the
+/// history is relevant for delta consumption, it is up to the observers to cache the
+/// history on their own.
+///
+/// While the protocol does not bind the `Changeset` associated type with any constraint,
+/// it is expected to be either a `Changeset`, or a composite of `Changeset`s for a nested
+/// collection.
+public struct Snapshot<Elements: Collection, Changeset>: SnapshotProtocol {
+	public let elements: Elements
+	public let changeset: Changeset
+
+	public init(elements: Elements, changeset: Changeset) {
+		(self.elements, self.changeset) = (elements, changeset)
+	}
+}
+
+public protocol SnapshotProtocol {
+	associatedtype Elements: Collection
+	associatedtype Changeset
+
+	var elements: Elements { get }
+	var changeset: Changeset { get }
+}
+
+/// Represents an atomic batch of changes made to a collection.
+///
+/// A `Changeset` represents changes as **offsets** of elements. You may
+/// subscript a collection of zero-based indexing with the offsets, e.g. `Array`. You must
+/// otherwise convert the offsets into indices before subscripting.
+public struct Changeset {
+	/// Represents the context of a move operation applied to a collection.
+	public struct Move {
+		public let source: Int
+		public let isMutated: Bool
+
+		public init(source: Int, isMutated: Bool) {
+			(self.source, self.isMutated) = (source, isMutated)
+		}
+	}
```

Comment:
This heterogeneous representation is not a not nice format to consume. There is an implicit order between all these operations, and say if you are interested in only a specific kind of operations at a time, you are forced to go over all of them again and again.

Moreover, these information can be (are) all prepared as standalone sets upfront.

Last but not least, while it has an memory overhead for small # of changes (4 references + 2 Ints per change), it is more compact when there is a lot of contiguous changes which `IndexSet` is optimised for.

https://github.com/ReactiveCocoa/ReactiveSwift/pull/494/files#diff-4a097c76f0bff02c4c19b2d40ecbc652R125